history/cycleValidator.go

```diff
@@ -0,0 +1,261 @@
+			eventType, err := event.EventType()
+			if len(sessionID) == 0 || err != nil {
+				continue
+			}
+
+			sessionIDs[sessionID] = true
+
+			if eventType == interpreter.OfflineEventType {
+				offlineEvents[sessionID] = true
+			}
+
+		}
+
+		if valid, invalidIds := determineError(sessionIDs, offlineEvents, excludeFunc); valid {
+			return true, nil
+		} else {
+			return false, CycleValidationErr{
+				OriginalErr:   ErrMissingOfflineEvent,
+				InvalidFields: invalidIds,
+				ErrorTag:      validation.MissingOfflineEvent,
+			}
+		}
+	}
+}
+
+func determineError(sessionIDs map[string]bool, existingEvents map[string]bool, exclude func(id string) bool) (bool, []string) {
+	var missingEvents []string
+	for id := range sessionIDs {
+		if !existingEvents[id] && !exclude(id) {
+			missingEvents = append(missingEvents, id)
+		}
+	}
+
+	if len(missingEvents) == 0 {
+		return true, nil
+	}
+
+	return false, missingEvents
+}
+
+func determineMetadataValues(fields []string, event interpreter.Event) map[string]string {
+	values := make(map[string]string)
+	for _, field := range fields {
+		if val, ok := event.Metadata[field]; ok {
+			values[field] = val
+		} else {
+			values[field] = ""
+		}
+	}
+
+	return values
+}
+
+func validateMetadata(keys []string, events []interpreter.Event) []string {
+	if len(events) == 0 {
+		return nil
+	}
+
+	metadataVals := determineMetadataValues(keys, events[0])
+	incorrectFieldsMap := make(map[string]bool)
+	for _, event := range events {
+		incorrectFieldsMap = checkMetadataValues(metadataVals, incorrectFieldsMap, event)
+	}
+
+	if len(incorrectFieldsMap) == 0 {
+		return nil
+	}
+
+	fields := make([]string, 0, len(incorrectFieldsMap))
+	for key := range incorrectFieldsMap {
+		fields = append(fields, key)
+	}
+
+	return fields
+
+}
+
+func validateMetadataWithinCycle(keys []string, events []interpreter.Event) []string {
+	if len(events) == 0 {
+		return nil
+	}
+
+	metadataVals := make(map[int64]map[string]string)
+	incorrectFieldsMap := make(map[string]bool)
+	for _, event := range events {
+		boottime, err := event.BootTime()
+		if err != nil || boottime <= 0 {
+			continue
+		}
+
+		expectedVals, found := metadataVals[boottime]
+		if !found {
+			expectedVals = determineMetadataValues(keys, event)
+			metadataVals[boottime] = expectedVals
+			continue
+		}
+
+		incorrectFieldsMap = checkMetadataValues(expectedVals, incorrectFieldsMap, event)
+	}
+
+	if len(incorrectFieldsMap) == 0 {
+		return nil
+	}
+
+	fields := make([]string, 0, len(incorrectFieldsMap))
+	for key := range incorrectFieldsMap {
+		fields = append(fields, key)
+	}
+
+	return fields
+
+}
+
+func checkMetadataValues(expectedMetadataVals map[string]string, incorrectMetadata map[string]bool, event interpreter.Event) map[string]bool {
+	for key, val := range expectedMetadataVals {
+		v, found := event.Metadata[key]
+		if found && v != val {
+			incorrectMetadata[key] = true
+		} else if !found && len(val) > 0 {
+			incorrectMetadata[key] = true
+		}
```

Comment:
As I commented above, when getting a value from a `map[string]string` where it's not found, the value returned is the empty string.  In this case, I don't really see the point in differentiating `!found` vs `v` being `""`? 

```suggestion
		if event.Metadata[key] != val {
			incorrectMetadata[key] = true
		}
```

Reply:
I could see checking `found` if we say that metadata with a key with an empty string value is different than not having a key...but then we would also need to change other logic to recognize these as different as well (it doesn't seem worth it to me).
```suggestion
		v, found := event.Metadata[key]
		if !found || v != val {
			incorrectMetadata[key] = true
		}
```